pkg/azure/provider/provider.go

```diff
@@ -0,0 +1,195 @@
+	factory access.Factory
+}
+
+// NewDefaultDriver creates a new instance of an implementation of provider.Driver. This can be mostly used by tests where we also wish to have our own polling intervals.
+func NewDefaultDriver(accessFactory access.Factory) driver.Driver {
+	return defaultDriver{
+		factory: accessFactory,
+	}
+}
+
+func (d defaultDriver) ListMachines(ctx context.Context, req *driver.ListMachinesRequest) (resp *driver.ListMachinesResponse, err error) {
+	defer instrument.RecordDriverAPIMetric(err, listMachinesOperationLabel, time.Now())
+	providerSpec, connectConfig, err := helpers.ExtractProviderSpecAndConnectConfig(req.MachineClass, req.Secret)
+	if err != nil {
+		return
+	}
+	vmNames, err := helpers.ExtractVMNamesFromVirtualMachinesAndNICs(ctx, d.factory, connectConfig, providerSpec.ResourceGroup, providerSpec.Tags)
+	if err != nil {
+		err = status.Error(codes.Internal, fmt.Sprintf("Failed to extract VM names from VMs and NICs for resourceGroup: %s, Err: %v", providerSpec.ResourceGroup, err))
+		return
+	}
+	resp = helpers.ConstructMachineListResponse(providerSpec.Location, vmNames)
+	return
+}
+
+func (d defaultDriver) CreateMachine(ctx context.Context, req *driver.CreateMachineRequest) (resp *driver.CreateMachineResponse, err error) {
+	defer instrument.RecordDriverAPIMetric(err, createMachineOperationLabel, time.Now())
+	providerSpec, connectConfig, err := helpers.ExtractProviderSpecAndConnectConfig(req.MachineClass, req.Secret)
+	if err != nil {
+		return
+	}
+	vmName := req.Machine.Name
+	nicName := utils.CreateNICName(vmName)
+
+	imageReference, plan, err := helpers.ProcessVMImageConfiguration(ctx, d.factory, connectConfig, providerSpec, vmName)
+	if err != nil {
+		return
+	}
+	subnet, err := helpers.GetSubnet(ctx, d.factory, connectConfig, providerSpec)
+	if err != nil {
+		return
+	}
+
+	nicID, err := helpers.CreateNICIfNotExists(ctx, d.factory, connectConfig, providerSpec, subnet, nicName)
+	if err != nil {
+		return
+	}
+
+	vm, err := helpers.CreateOrUpdateVM(ctx, d.factory, connectConfig, providerSpec, imageReference, plan, req.Secret, nicID, vmName)
+	if err != nil {
+		return
+	}
+	resp = helpers.ConstructCreateMachineResponse(providerSpec.Location, vmName)
+	helpers.LogVMCreation(providerSpec.Location, providerSpec.ResourceGroup, vm)
+	return
+}
+
+func (d defaultDriver) DeleteMachine(ctx context.Context, req *driver.DeleteMachineRequest) (resp *driver.DeleteMachineResponse, err error) {
+	defer instrument.RecordDriverAPIMetric(err, deleteMachineOperationLabel, time.Now())
+	providerSpec, connectConfig, err := helpers.ExtractProviderSpecAndConnectConfig(req.MachineClass, req.Secret)
+	if err != nil {
+		return
+	}
+	var (
+		resourceGroup = providerSpec.ResourceGroup
+		vmName        = strings.ToLower(req.Machine.Name)
+	)
+	// Check if Deletion of the machine (VM, NIC, Disks) can be completely skipped.
+	skipDelete, err := helpers.SkipDeleteMachine(ctx, d.factory, connectConfig, resourceGroup)
```

Comment:
Could you clarify , if a resource group doesn't exist, why a DeleteMachine call be done? or is the intent to make DeleteMachine deal with all corner cases on being called separately?

Reply:
It is a driver implementation and does not assume the correctness of the caller nor does the API documentation state that its expected that the resource group should exist when invoking this method. Since the Driver interface is provider agnostic therefore resource group existence cannot be put as a generic requirement on the API consumers since that is specific to azure.